src/main/kotlin/de/bund/digitalservice/useid/widget/WidgetController.kt

```diff
@@ -127,4 +143,24 @@ class WidgetController(
     private fun setEiDClientURL(url: String): Pair<String, String> {
         return "eidClientURL" to url
     }
+
+    /*
+       Since user agent will return version in String, we need to properly convert the string to integer,
+       otherwise we cannot perform logical operations such as <= or >=
+       If major version returns empty or somehow the user agent parser cannot parse the version properly,
+       we will show to users the widget
+    */
+    private fun safelyGetMajorOSVersion(majorVersion: String?): Int? {
+        return try {
+            Integer.parseInt(majorVersion)
+        } catch (exception: NumberFormatException) {
+            null
+        }
+    }
+
+    private fun hasSupportedMajorOSVersion(parsedUserAgent: Client, osFamily: String, supportedMajorVersion: Int): Boolean {
+        return parsedUserAgent.os.family == osFamily &&
+            parsedUserAgent.os.major != null &&
+            safelyGetMajorOSVersion(parsedUserAgent.os.major)!! <= supportedMajorVersion
```

Comment:
If `safelyGetMajorOSVersion` returns `null` this will throw an exception which will lead to an 500 error response to the user. This must be handled gracefully.

Reply:
so, `parsedUserAgent.os.major` will always return value of type `String` in case the the user agent is correct or wrong (empty string for invalid or missing).  
Therefor I simplify the logic for checking the major version, also I will remove `safelyGetMajorOSVersion` which I will explain at the bottom.

```kotlin
private fun hasIncompatibleMajorVersion(parsedUserAgent: Client, osFamily: String, supportedMajorVersion: Int): Boolean {
        return parsedUserAgent.os.family == osFamily &&
            !parsedUserAgent.os.major.isNullOrEmpty() &&
            Integer.parseInt(parsedUserAgent.os.major) <= supportedMajorVersion
    }
```

since we are using double ampersand `&&` operator, if the first left statement false, it won't evaluate the right side (short-circuit evaluation).

```kotlin
!parsedUserAgent.os.major.isNullOrEmpty() &&
            Integer.parseInt(parsedUserAgent.os.major) <= supportedMajorVersion
```
if `!parsedUserAgent.os.major.isNullOrEmpty()` returns true (e.g not null, not empty or has value) then it will evaluate the second part `Integer.parseInt(parsedUserAgent.os.major) <= supportedMajorVersion` with the existing value from left part

if `!parsedUserAgent.os.major.isNullOrEmpty()` returns false (e.g null, empty or no value) then it will short-circuit and won't evaluate the second part (it won't call the function  like following `Integer.parseInt("")`) which will not throw `NumberFormatException`

what do you think?

Reply:
Reasonable. I would still wrap the whole user-agent parsing in a try-catch block since this should not influence the response to the user in case of any runtime exceptions. In case of any exception the widget is just returned as usual.